### PR TITLE
Application insights updates

### LINKED ISSUE #9254: Azure AD authentication for Application Insights is not working in Azure Function
I am getting the below error when I am trying to use Azure AD authentication for Application Insights in my Azure Function.
`Microsoft.ApplicationInsights: The provided tokenCredential must inherit Azure.Core.TokenCredential (Parameter 'tokenCredential').`
I am getting this error both in my local machine and also in my azure portal as a Functions runtime error.

#### Investigative information

- Timestamp: 2023-05-04T08:00:00Z
- Function App version: 4
- Function App name: func-tpgcp-mail-dev-eu
- Region: North Europe

#### Repro steps

Here is the code that I wrote in my Startup.Configure:
```
var managedIdentityCredential = new ManagedIdentityCredential();
services.Configure<TelemetryConfiguration>(config => config.SetAzureTokenCredential(managedIdentityCredential));
services.AddApplicationInsightsTelemetryWorkerService();
```

#### Expected behavior

The function app can use its managed identity to authenticate and sends telemetry data to application insights.

#### Actual behavior

As I said in the description I am getting an error that says `ManagedIdentityCredential` is not inherited from the `TokenCredential` class.

#### Known workarounds

Nothing came to my mind.

#### Related information 

After some investigation, I realized that `ReflectionCredentialEnvelope` uses reflection to load the `TokenCredential` type and then uses the type to check if the passed credential object inherits from that type or not. Then I took a look at the loaded modules in the application on my local machine and saw that the `Azure.Core.dll` is loaded twice with different versions:

image

Also here are all packages that I used in my app function:

```
<ItemGroup>
    <PackageReference Include="Azure.Core" Version="1.31.0" />
    <PackageReference Include="Azure.Identity" Version="1.8.2" />
    <PackageReference Include="Azure.Messaging.ServiceBus" Version="7.13.1" />
    <PackageReference Include="Azure.Storage.Blobs" Version="12.16.0" />
    <PackageReference Include="Microsoft.ApplicationInsights.WorkerService" Version="2.21.0" />
    <PackageReference Include="Microsoft.Azure.Cosmos" Version="3.33.0" />
    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.1.0" />
    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.CosmosDB" Version="4.2.0" />
    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.EventGrid" Version="3.2.1" />
    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.ServiceBus" Version="5.9.0" />
    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.Storage" Version="5.1.2" />
    <PackageReference Include="Microsoft.Graph" Version="5.8.0" />
    <PackageReference Include="Microsoft.Identity.Client" Version="4.53.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="4.2.0" />
  </ItemGroup>
```

### PR DESCRIPTION
**Pull request by RohitRanjanMS**
1. [IMPORTANT] Updating Azure.Core->1.36.0 and Azure.identity->1.10.4 to align with [PR](https://github.com/Azure/azure-webjobs-sdk/pull/3046).
2. Updating Microsoft.Azure.WebJobs.Logging.ApplicationInsights.
Enabling support for-
     - ManagedIdentity
     - AutocollectedMetricsExtractor
     - QueryStringTracing
     - MetricsCustomDimensionOptimization
     - ApplicationInsights Event logging
     - ApplicationInsights Log Diagnostics
3. Updating ApplicationInsights related packages to 2.22.
4. Updating ScriptHostBuilder to use auth string and log level.
5. Updating RecordAndThrowExternalStartupException to use auth string.

I will update the doc after the release.
### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj [hunk 1]
```diff
@@ -62,13 +62,13 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Azure.Identity" Version="1.10.2" />
+    <PackageReference Include="Azure.Identity" Version="1.10.4" />
```
Comment (jviau) [inline src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj:65]: @brettsam @fabiocav are we okay with this change? Did we verify that downstream major version rev is okay?

Comment (brettsam) [reply src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj:65]: This is a patch version update isn't it? If so it should be fine...

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 2]
```diff
@@ -408,10 +415,30 @@ internal static void ConfigureApplicationInsights(HostBuilderContext context, IL
...
+                        }
+                        else
+                        {
+                            throw new InvalidEnumArgumentException($"Invalid `{EnvironmentSettingNames.AppInsightsEventListenerLogLevel}`.");
+                        }
+                    }
+
+                    if (!string.IsNullOrEmpty(context.Configuration[EnvironmentSettingNames.AppInsightsAuthenticationString]))
+                    {
+                        o.TokenCredentialOptions = TokenCredentialOptions.ParseAuthenticationString(context.Configuration[EnvironmentSettingNames.AppInsightsAuthenticationString]);
```
Comment (brettsam) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs:440]: 2 questions here:
- What are the requirements to use this? An Instrumentation key *or* a Conn string are required? 
- I see this parsing is in our own code in WebJobs -- are there docs on what this auth string is meant to look like? is it readily copy-able from somewhere in the portal?

File: src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj (hunk 1 above)
Comment (jviau) [reply src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj:65]: This is the one which downstream brings in that major version rev we were concerned about before. Not sure if we resolved that and determined the update was safe.

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs (hunk 2 above)
Comment (RohitRanjanMS) [reply src/WebJobs.Script/ScriptHostBuilderExtensions.cs:440]: Yes, need either an IKey or connection string. 

[APPLICATIONINSIGHTS_AUTHENTICATION_STRING](https://msazure.visualstudio.com/One/_git/Mgmt-AppInsights-StatusMonitor?path=%2FAgents%2FCommon%2FApplicationInsights.Identity%2Freadme.md&_a=contents&version=GBmaster#) is owned by the ApplicationInsights team. I have used some of their logic/code for consistency (like parsing, validation and error scenarios). 

I will update our documentation on how to enable AAD auth, including details about the auth string.

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 3]
```diff
@@ -408,10 +415,30 @@ internal static void ConfigureApplicationInsights(HostBuilderContext context, IL
...
             {
+                // Initialize ScriptTelemetryInitializer before any other telemetry initializers.
+                // This will allow HostInstanceId to be removed as part of MetricsCustomDimensionOptimization.
+                builder.Services.AddSingleton<ITelemetryInitializer, ScriptTelemetryInitializer>();
                 builder.AddApplicationInsightsWebJobs(o =>
                 {
                     o.InstrumentationKey = appInsightsInstrumentationKey;
                     o.ConnectionString = appInsightsConnectionString;
+
+                    if (!string.IsNullOrEmpty(context.Configuration[EnvironmentSettingNames.AppInsightsEventListenerLogLevel]))
```
Comment (jviau) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs:426]: small nit for these double indexes into IConfiguration. Can use pattern matching:

``` CSHarp
if (context.Configuration[<Key>] is string value && value is not "")
{
    // use 'value' here
}
```

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 4]
```diff
@@ -261,7 +267,8 @@ public static IHostBuilder AddScriptHostCore(this IHostBuilder builder, ScriptAp
                         {
                             // Go directly to the environment; We have no valid configuration from the customer at this point.
                             string appInsightsConnStr = Environment.GetEnvironmentVariable(EnvironmentSettingNames.AppInsightsConnectionString);
-                            RecordAndThrowExternalStartupException("Error building configuration in an external startup class.", ex, loggerFactory, appInsightsConnStr);
+                            string appInsightsAuthStr = Environment.GetEnvironmentVariable(EnvironmentSettingNames.AppInsightsAuthenticationString);
```
Comment (jviau) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs:270]: This logic of checking two different ways of authing to app insights is repeated a few times. Is there a way we can extract into a common helper? Maybe one that optionally takes in `IConfiguration`, will try to extract the two values, then fallsback to `Environment` if `IConfiguration` was not provided, or it did not contain the necessary auth info.

**Review by jviau — APPROVED** [review]
Looks good from my perspective.

File: test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json [hunk 5]
```diff
@@ -6,27 +6,27 @@
   "compilationOptions": {},
```
Comment (brettsam) [inline test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json:6]: Can you back this file out, run the test, and give us the output (as a comment)? I see a couple of concerning major version bumps there and want to see them all summarized.

Comment (RohitRanjanMS) [reply test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json:6]: sure

Comment (RohitRanjanMS) [reply test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json:6]: Message: 
IMPORTANT: The dependencies in WebHost have changed and MUST be reviewed before proceeding. Please follow up with brettsam, fabiocav or mathewc for approval.

Previous file: C:\Users\Rohit Ranjan\source\repos\azure-functions-host\test\WebJobs.Script.Tests\bin\Debug\net6.0\Microsoft.Azure.WebJobs.Script.WebHost.deps.json
New file:      C:\Users\Rohit Ranjan\source\repos\azure-functions-host\src\WebJobs.Script.WebHost\bin\Debug\net6.0\Microsoft.Azure.WebJobs.Script.WebHost.deps.json

  Changed:
    - Azure.Core.dll: 1.35.0.0/1.3500.23.45706 -> 1.36.0.0/1.3600.23.56006
    - Azure.Identity.dll: 1.10.2.0/1.1000.223.50903 -> 1.10.4.0/1.1000.423.56303
    - Microsoft.AI.DependencyCollector.dll: 2.21.0.429/2.21.0.429 -> 2.22.0.997/2.22.0.997
    - Microsoft.AI.EventCounterCollector.dll: 2.21.0.429/2.21.0.429 -> 2.22.0.997/2.22.0.997
    - Microsoft.AI.PerfCounterCollector.dll: 2.21.0.429/2.21.0.429 -> 2.22.0.997/2.22.0.997
    - Microsoft.AI.ServerTelemetryChannel.dll: 2.21.0.429/2.21.0.429 -> 2.22.0.997/2.22.0.997
    - Microsoft.AI.WindowsServer.dll: 2.21.0.429/2.21.0.429 -> 2.22.0.997/2.22.0.997
    - Microsoft.ApplicationInsights.AspNetCore.dll: 2.21.0.429/2.21.0.429 -> 2.22.0.997/2.22.0.997
    - Microsoft.ApplicationInsights.dll: 2.21.0.429/2.21.0.429 -> 2.22.0.997/2.22.0.997
    - Microsoft.ApplicationInsights.SnapshotCollector.dll: 1.4.3.0/1.4.3.0 -> 1.4.4.0/1.4.4.0
    - Microsoft.Extensions.Logging.ApplicationInsights.dll: 2.21.0.429/2.21.0.429 -> 2.22.0.997/2.22.0.997
    - Microsoft.Identity.Client.dll: 4.54.1.0/4.54.1.0 -> 4.56.0.0/4.56.0.0
    - Microsoft.Identity.Client.Extensions.Msal.dll: 2.31.0.0/2.31.0.0 -> 4.56.0.0/4.56.0.0
    - Microsoft.IdentityModel.Abstractions.dll: 6.35.0.0/6.35.0.41201 -> 6.32.0.0/6.32.0.40712
    - Microsoft.IdentityModel.JsonWebTokens.dll: 6.35.0.0/6.35.0.41201 -> 6.32.0.0/6.32.0.40712
    - Microsoft.IdentityModel.Logging.dll: 6.35.0.0/6.35.0.41201 -> 6.32.0.0/6.32.0.40712
    - Microsoft.IdentityModel.Protocols.dll: 6.35.0.0/6.35.0.41201 -> 6.32.0.0/6.32.0.40712
    - Microsoft.IdentityModel.Protocols.OpenIdConnect.dll: 6.35.0.0/6.35.0.41201 -> 6.32.0.0/6.32.0.40712
    - Microsoft.IdentityModel.Tokens.dll: 6.35.0.0/6.35.0.41201 -> 6.32.0.0/6.32.0.40712
    - Microsoft.Win32.SystemEvents.dll: 4.0.2.0/4.700.19.56404 -> 6.0.0.0/6.0.21.52210
    - System.Configuration.ConfigurationManager.dll: 4.0.3.0/4.700.19.56404 -> 6.0.0.0/6.0.21.52210
    - System.Diagnostics.PerformanceCounter.dll: 4.0.2.0/4.700.19.56404 -> 6.0.0.0/6.0.21.52210
    - System.Drawing.Common.dll: 4.0.2.3/4.700.21.51508 -> 6.0.0.0/6.0.21.52210
    - System.IdentityModel.Tokens.Jwt.dll: 6.35.0.0/6.35.0.41201 -> 6.32.0.0/6.32.0.40712
    - System.Security.Cryptography.ProtectedData.dll: 4.0.5.0/4.700.19.56404 -> 6.0.0.0/6.0.21.52210
    - System.Security.Permissions.dll: 4.0.3.0/4.700.19.56404 -> 6.0.0.0/6.0.21.52210
    - System.Windows.Extensions.dll: 4.0.1.0/4.700.19.56404 -> 6.0.0.0/6.0.21.52210

  Removed:

  Added:

Expected: True
Actual:   False

Comment (RohitRanjanMS) [reply test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json:6]: @brettsam

Comment (brettsam) [reply test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json:6]: Well... I'm not sure we can do any of this without a bigger review. That's a lot of major version bumps. 

- Microsoft.Identity.Client.Extensions.Msal.dll: 2.31.0.0/2.31.0.0 -> 4.56.0.0/4.56.0.0
- Microsoft.Win32.SystemEvents.dll: 4.0.2.0/4.700.19.56404 -> 6.0.0.0/6.0.21.52210
- System.Configuration.ConfigurationManager.dll: 4.0.3.0/4.700.19.56404 -> 6.0.0.0/6.0.21.52210
- System.Security.Cryptography.ProtectedData.dll: 4.0.5.0/4.700.19.56404 -> 6.0.0.0/6.0.21.52210
- System.Security.Permissions.dll: 4.0.3.0/4.700.19.56404 -> 6.0.0.0/6.0.21.52210
- System.Windows.Extensions.dll: 4.0.1.0/4.700.19.56404 -> 6.0.0.0/6.0.21.52210

**Comment by RohitRanjanMS** [comment]
Closing this in favor of https://github.com/Azure/azure-functions-host/pull/9835.

**Pull request closed without merging**